### PR TITLE
test: refactor unit tests with solc downloader

### PR DESCRIPTION
**Pull request by antonbaliasnikov**
# What ❔

* [x] Moved `src/tests` to `tests` folder for more canonical tests organization accordingly to [the Rustbook](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests)
* [x] Add `Downloader` module in tests that downloads all required `solc` versions on-the-fly
* [x] Use `zksolc` with help of [`assert_cmd` crate](https://docs.rs/assert_cmd/latest/assert_cmd/) instead of requirement of adding it to the `PATH`
* [x] Small refactoring and unification of testing code  

## Why ❔

To allow running regression tests using `cargo test` only without putting anything to the `PATH` before.

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [x] Code has been formatted via `cargo fmt` and checked with `cargo clippy`.

### CODE REVIEW

File: tests/common/downloader/mod.rs [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+///
+#[derive(Debug)]
+pub struct Downloader {
+    /// The `reqwest` HTTP client.
+    http_client: reqwest::blocking::Client,
+    /// The solc-bin JSON list metadata.
+    solc_list: Option<SolcList>,
+}
+
+impl Downloader {
```
Comment (antonbaliasnikov) [inline tests/common/downloader/mod.rs:30]: @hedgar2017 , I've decided to copy implementation of the `Downloader` from the `compiler-tester` here and have a similar `solc-bin.json` config for it.

I was thinking of two other solutions:
1. Greatly simplify `Downloader` to just downloaded hardcoded versions from the tests
  * I was thinking that we probably want to be agile and support the same config as in the tester? 
3. Re-use `Downloader` from the `compiler-tester`.
  * The argument to not go this way was not to have `tester` as a dependency for FE compiler (even for tests)

I'm not certain in this solution, so, need your advice here. What do you think would be the best for us here?

Comment (hedgar2017) [reply tests/common/downloader/mod.rs:32]: Let's convert `era-compiler-solidity` to a workspace and make `downloader` a separate crate.
We'll need the workspace for @sayon's Yul parser anyway.

File: tests/common/mod.rs [hunk 2]
```diff
@@ -4,30 +4,91 @@
...
 use std::collections::BTreeMap;
 use std::collections::BTreeSet;
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
 use std::str::FromStr;
+use std::sync::Once;
+use std::time::Duration;
+
+/// Synchronization for solc downloads.
+#[cfg(target_family = "unix")]
```
Comment (hedgar2017) [inline tests/common/mod.rs:28]: What do we do on Windows?

**Review by hedgar2017 — CHANGES_REQUESTED** [review]
Nice! Just a nits or two.

Are they passing already?
Is most of this translated by AI? :)

File: tests/common/downloader/mod.rs (hunk 1 above)
Comment (hedgar2017) [reply tests/common/downloader/mod.rs:32]: But the problem is that the Downloader is also used for vyper, and we can't have the same code in both crates.
Maybe worth doing it in `era-compiler-common` by converting it into a workspace? The repo was made for common stuff not depending on LLVM.

File: tests/common/mod.rs [hunk 3]
```diff
@@ -4,30 +4,82 @@
...
 use std::collections::BTreeMap;
 use std::collections::BTreeSet;
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
 use std::str::FromStr;
+use std::sync::Once;
+use std::time::Duration;
+
+/// Synchronization for solc downloads.
+static DOWNLOAD_SOLC: Once = Once::new();
```
[the remark this answers is not in the document]
Comment (antonbaliasnikov) [reply tests/common/mod.rs:29]: Sure, I'm adding comments in the source code. Opened a discussion just to confirm with you as don't have too much experience in this. Thanks, @hedgar2017 !

File: tests/common/downloader/mod.rs (hunk 1 above)
Comment (antonbaliasnikov) [reply tests/common/downloader/mod.rs:32]: That makes total sense, I'll put downloaded to `era-compiler-common`. Thanks!

File: tests/common/mod.rs (hunk 2 above)
Comment (antonbaliasnikov) [reply tests/common/mod.rs:28]: Today I tried to make it work for Windows, the main thing that is missing for it it's an older version of `solc` that is downloaded from Solidity mirror, and they are in `zip` archives with additional `dll` libraries... So, for now, on Windows, I'm just using my action that downloads `solc` and unpacks everything.

I'll try to add a proper Windows support in `era-compiler-common`.

**Comment by antonbaliasnikov** [comment]
> Nice! Just a nits or two.
> 
> Are they passing already? Is most of this translated by AI? :)

Yeah, I've used ChatGPT/Copilot providing more and more examples for them, but they still make lots of mistakes in Rust. ChatGPT is unable to work properly with ownership, and always making mistakes due to outdated library versions that he knows.

So, even with AI, there's a lot of work to do... I hope it will get better!

**Review by hedgar2017 — APPROVED** [review]

**Comment by antonbaliasnikov** [comment]
@hedgar2017 , thanks for the review!
I updated the `era-compiler-common` branch after our merge, ready to be merged now.

**Pull request merged**